### PR TITLE
ARROW-7745: [Doc] [C++] Update Parquet documentation

### PR DESCRIPTION
**Pull request by gawain-bolton**
Add documentation for new StreamReader and StreamWriter classes.

Also document ParquetException class and Parquet PARQUET_THROW_NOT_OK
and PARQUET_ASSIGN_OR_THROW macros.

### CODE REVIEW

File: docs/source/cpp/parquet.rst [hunk 1]
```diff
@@ -27,11 +27,25 @@ Reading and writing Parquet files
...
 from tight integration with Arrow C++.
 
-Reading
-=======
+The :class:`arrow::FileReader` class reads data for an entire
+file or row group into an :class:`::arrow::Table`.
 
-The Parquet :class:`FileReader` requires a :class:`::arrow::io::RandomAccessFile`
-instance representing the input file.
+The :func:`arrow::WriteTable` method writes an entire
```
Comment (pitrou) [inline docs/source/cpp/parquet.rst]: "function" rather than "method", no?

File: docs/source/cpp/parquet.rst [hunk 2]
```diff
@@ -27,11 +27,25 @@ Reading and writing Parquet files
...
+:class:`::arrow::Table` to an output file.
+
+The :class:`StreamReader` and :class:`StreamWriter` classes allow for
+data to be written using a C++ input/output streams approach to
+read/write fields column by column and row by row.  This is approach
```
Comment (pitrou) [inline docs/source/cpp/parquet.rst]: "This approach"

File: docs/source/cpp/parquet.rst [hunk 3]
```diff
@@ -27,11 +27,25 @@ Reading and writing Parquet files
...
+is offered for ease of use and type-safety.  It is of course also
+useful when data must be streamed as files are read and written
+incrementally.
```
Comment (pitrou) [inline docs/source/cpp/parquet.rst]: Should you add a word about performance? I presume it's not great compared to other IO facilities.

File: docs/source/cpp/parquet.rst [hunk 4]
```diff
@@ -58,12 +72,119 @@ instance representing the input file.
...
+:class:`arrow::FileReaderBuilder` helper class.
 
 .. TODO write section about performance and memory efficiency
 
-Writing
-=======
+WriteTable
+==========
+
+The :func:`arrow::WriteTable` method writes an entire
```
Comment (pitrou) [inline docs/source/cpp/parquet.rst]: "function" perhaps?

**Review by pitrou — COMMENTED** [review]
LGTM, just a couple nits

**Comment by emkornfield** [comment]
@gawain-bolton I don't think the failures are real, would you mind rebasing and seeing if they go away?

**Pull request closed without merging**